### DISCUSSION TITLE
Compose: Small Stuff

### DISCUSSION THREAD

**Discussion by flatcap** [neomutt/neomutt#3222] [Development]
Here are a bunch of small(-ish) ideas for Compose.
In no particular order...

### Functions: `<move-to-top>`, `<move-to-bottom>`

Early in development, we had these functions to move an attachment to the top/bottom (of its group/level).

Do we need them?
Might they be useful?

### Functions: `<move-out-of-group>`, `<move-into-group>`

These are effectively, "move-left" and "move-right".

`<move-into-group>` would promote an attachment into the group immediately preceding it.
It would be subject to the same rules applied when making that type of group.

`<move-out-of-group>` would simply move the attachment up one level and place it immediately after the group it was in.

### Translations

We need translations for:
- "Related parts for"
- "Multilingual part for"
- "Alternatives for"

On a related note, do we want to **send** these strings?
Are they useful for the composer, the receiver, or both, or neither?

If they're only useful for the composer, could we "fake" them?
Display something like:

> [Related parts]
> [Related parts for file]

### Expandos

The composer might want to see the `Content-ID`, if it's autogenerated.
We could create an expando (`%I`) for it.

Also, make sure every attribute of the attachment has an expando.

### Patterns

Create some new Patterns, so that the user can search for any of the attachment's attributes, e.g. `~I content-id`

### Colours

Introduce colour objects for common attachment properties,
e.g.
- Attached vs Inline
- Signed / Encrypted
- Group Related / Alternatives / Lingual

Also allow regex pattern matching for custom colouring.
e.g.
- `color attachment red default '~T image/'`

### Bug: Collapsed Grouped Attachment

When viewing received attachments, the user can collapse groups.

**Before** folding:
```
1  <no description>
2  alternates.c
3  Alternatives for "alternates.h"
4  ├─>alternates.h
5  └─>command_parse.c
6  command_parse.h
```

**After** folding:
```
1  <no description>
2  alternates.c
3  Alternatives for "alternates.h"
4  command_parse.h
```

The attachments have been renumbered, so there's no evidence that group contents actually exist.

Displaying the absolute numbers would help.
There's a pattern for a collapsed Email Thread and an expando for the number of hidden Emails.
We could do the same and put it in the default `$attach_format`.

e.g.
```
3  [2] Alternatives for "alternates.h"
```

### Insert Attachments at the Cursor

When attaching files, they always appear at the end of the list.

We could change the behaviour to add at the cursor's position.
(change this behaviour before too many people start using the new code)

The big advantage of this, is that the user can **add** to an existing group.
Currently, they'd have to:
- Attach file
- Break group
- Shuffle attachments
- Re-create group

**Comment by dcpurton** [comment]
> ### Translations
> We need translations for:
> 
> * "Related parts for"
> * "Multilingual part for"
> * "Alternatives for"
> 
> On a related note, do we want to **send** these strings? Are they useful for the composer, the receiver, or both, or neither?
> 
> If they're only useful for the composer, could we "fake" them? Display something like:
> 
> > [Related parts]
> > [Related parts for file]

I don't think these descriptions are particularly useful at all.

I removed them by default in a number of situations. But I can't really see why they couldn't be removed completely. If the user really wants one, they could still add one.

**Reply by flatcap** [reply]
OK, then they can go.

**Reply by dcpurton** [reply]
See PR #3226

**Comment by dcpurton** [comment]
> ### Bug: Collapsed Grouped Attachment
> When viewing received attachments, the user can collapse groups.
> 
> **Before** folding:
> 
> ```
> 1  <no description>
> 2  alternates.c
> 3  Alternatives for "alternates.h"
> 4  ├─>alternates.h
> 5  └─>command_parse.c
> 6  command_parse.h
> ```
> 
> **After** folding:
> 
> ```
> 1  <no description>
> 2  alternates.c
> 3  Alternatives for "alternates.h"
> 4  command_parse.h
> ```
> 
> The attachments have been renumbered, so there's no evidence that group contents actually exist.
> 
> Displaying the absolute numbers would help. There's a pattern for a collapsed Email Thread and an expando for the number of hidden Emails. We could do the same and put it in the default `$attach_format`.
> 
> e.g.
> 
> ```
> 3  [2] Alternatives for "alternates.h"
> ```

Would it make sense to treat collapsed groups the same way as collapsed threads?

Do collapsed email threads always have discontinuous numbering? Or is that a setting?

So, it could be nice to be able to:

- colour groups differently depending on whether they are collapsed or not
- format them using something like is done for emails in the `index` (this allows to show number of emails in thread, right?) So you could show number of attachments in group.

**Reply by flatcap** [reply]
> Would it make sense to treat collapsed groups the same way as collapsed threads?

Yes

> Do collapsed email threads always have discontinuous numbering?

Always

> colours

Yes
Can you make a list of everything (attachment-y) that we might want to colour

> formatting

Yes
What extra expandos do we need?

**Reply by dcpurton** [reply]
I'm not as familiar with Neomutt's formatting and colour options. But will take a look. Might be a good incentive to improve my own view, which I just pulled of some random website…

**Reply by flatcap** [reply]
The colours work in two ways:
- Objects: Define some "things" that are coloured
  e.g. `[multipart/related]` group
- Regex: Colour any line, or partial line, by matching a regex against some attribute of the attachment
  e.g. `color attach '~t text/html' green default`

**Comment by dcpurton** [comment]
To be honest, I went through so much pain with the Body pointers, that I'm not sure it's worth adding things like `<move-out-of-group>`, `<move-into-group>`, `<move-to-top>`, `<move-to-bottom>`… At least not until you get a feature request from someone who shows that it's worth doing.

It was necessary to make sure that existing commands like `<move-up>` and `<move-down>` worked. But even these will probably not normally be used. Most email structures are not complicated.

In theory `<insert-file>` should straight forward, since the grouping code already does this when it inserts the group.

There's also `<new-mime>` which would need a corresponding `<insert-mime>`.

One important reason why you need both `<attach-file>` and `<insert-file>` is that you must have a way of adding an attachment after the current cursor position. Perhaps more useful descriptive terms would be `<attach-file-after>` and `<attach-file-before>`. You could then either replace `<attach-file>` with `<attach-file-after>` or just have all three and change the default mapping of `a` to `<attach-file-after>`. Though this would break current macros I suspect.

Having `<attach-file-before>` and `<attach-file-after>` also reduces the need for `<move-into-group>` and `<move-out-of-group>`.

**Comment by flatcap** [comment]
#3223 Support for inserting attachments

**Comment by flatcap** [comment]
> I went through so much pain with the Body pointers

It's really appreciated :-)

> that I'm not sure it's worth adding things like `<move-out-of-group>`, `<move-into-group>`

I _do_ think they're worth it (for a full suite of motion commands), but not yet.
I'm not fussed about the top/bottom commands.

`struct Body` **really** needs sorting out.
It'll be a LOT of work, but now seems the best time to tackle it
(as we have a reasonable grasp of how things fit together).

**Reply by dcpurton** [reply]
Yeah, having done the `<attach-file-*>` code I think other things are not so bad now. I can write the code to promote and demote attachments without too much trouble.

**Comment by flatcap** [comment]
As for inserting attachments, I think separate commands is going to make things too complicated.
You've covered `<attach-file>`, but there's also `<attach-message>` and `<attach-key>`.

You were concerned that:  "attach == insert-after-cursor" would break people's macros.
I agree, but I think we should focus on a simple interface.
The average user wouldn't see any change: after attaching a file, the cursor ends up on the last one.

As for the early-adopters / power-users -- I'll apologise to each of them personally :-)

I'm still waiting to see a list of use-cases for all this new code.
A list of common email/attachment layouts and all the steps necessary for generating them.

**Reply by dcpurton** [reply]
I thought about `<attach-message>` and `<attach-key>`. But I can not think of a reason for these not to be attached at the top level. (Not that it's a problem to code them.)

As for whether to have three commands. Unless I miss something obvious, this is useful.

My present code for `<attach-file-after>` and `<attach-file-before>` attaches the file at the same level as the currently selected attachment.

But if you just want to attach a file at the top level at the bottom of the list this becomes awkward as soon as your message ends with a group. You have to move the cursor to the last group at the top level and then `<attach-file-after>`. You can't know where this is, so it becomes impossible to automate.

You have the same problem with `<attach-file-before>`. If you remove this, inserting an attachment at the start of a group becomes impossible.

---

I guess, one way you *could* simplify things is to return to just *one* command, which always adds the attachment as the last one if a top level attachment is selected, or the last one in the current group if a grouped attachment is selected. This would have two advantages:

1. It would not break existing macros
2. It would allow adding attachments to groups in an exactly analogous way as attachments are currently added to a message.

**Reply by dcpurton** [reply]
> I'm still waiting to see a list of use-cases for all this new code. A list of common email/attachment layouts and all the steps necessary for generating them.

How do you want to do this and what do you have in mind?

**Reply by dcpurton** [reply]
> You were concerned that: "attach == insert-after-cursor" would break people's macros. I agree, but I think we should focus on a simple interface. The average user wouldn't see any change: after attaching a file, the cursor ends up on the last one.

This is *sort of* true. If you attach multiple files, each one is inserted at `idxlen` resulting in the order of attached files being reversed and the cursor position remaining at `idxlen`. So the cursor position is on the last attached file, but this isn't the bottom of the list.

To my mind this behaviour is odd, but I didn't change it for the new insert code I wrote. I would think that it would be better if attaching multiple files at once produced the same result as attaching the same files one at a time.

**Reply by flatcap** [reply]
> Inserting before group
> Appending after group

Hmm... good point.

**Reply by flatcap** [reply]
> > list of use-cases

> what do you have in mind

Start a new Discussion, with some example emails.
Several different formats of emails that we think might be desired.

Cut'n'paste the tree layout from compose,
and describe any pre-processing steps that are needed to generate a valid email.
